Add String#byteindex and String#byterindex

choznerol opened the pull request:
Fix #2361 

In this PR, `byteindex` and `byterindex` will become available for `String` just like `index` and `rindex`.

<img width="823" alt="image" src="https://user-images.githubusercontent.com/12410942/226643376-80f65298-4cfb-474f-a878-116a1e8a26bb.png">

The incorrect output of `index` and `rindex` is a separate bug that is tracked by #2360.

Please let me know if equivalent amount of doctest is preferred for all four methods. I was just a bit of worry about being too verbose, so there isn't too many test case for some of them.

choznerol committed:
Test String::index,rindex and add FIXME for #2360

choznerol committed:
Add byteindex and byterindex methods to spinoso-string

Close #2361

choznerol committed:
Expose byterindex just like byteindex

choznerol committed:
Use byteindex instead of index

choznerol commented:
> thread 'rustc' panicked at 'no resolutions for a doc link', compiler/rustc_metadata/src/rmeta/encoder.rs:2235:18

The [current CI failure](https://github.com/artichoke/artichoke/actions/runs/4480445128/jobs/7875764994?pr=2449) seems to be related to https://github.com/rust-lang/rust/issues/109424

lopopolo commented on artichoke-backend/src/extn/core/string/trampoline.rs:
```diff
@@ -601,7 +601,32 @@ pub fn byteindex(
+) -> Result<Value, Error> {
+    #[cfg(feature = "core-regexp")]
+    if let Ok(_pattern) = unsafe { Regexp::unbox_from_value(&mut substring, interp) } {
+        return Err(NotImplementedError::from("String#byteindex with Regexp pattern").into());
```
It looks like this error message should say `String#byterindex`.

choznerol replied:
opps thanks, fixed in 2b7662f0d0d1c4bf695e9f2262af61f17da26f22

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1635,11 +1635,13 @@ impl String {
-    /// assert_eq!(s.index("lo", None), Some(3));
-    /// assert_eq!(s.index("a", None), None);
-    /// assert_eq!(s.index("l", Some(3)), Some(3));
+    /// let s = String::from("via 💎 v3.2.0");
```
Can we use the `String::utf8` constructor here for clarity?

choznerol replied:
Sure, fixed in 44a6a075737f9c176f3c846ac642eb87de1b9f76

lopopolo commented on spinoso-string/src/lib.rs:
````diff
@@ -1666,6 +1668,25 @@ impl String {
+    /// ```
+    /// use spinoso_string::String;
+    ///
+    /// let s = String::from("via 💎 v3.2.0");
````
Ditto

choznerol replied:
ditto 44a6a075737f9c176f3c846ac642eb87de1b9f76

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1666,6 +1668,25 @@ impl String {
+    /// `String`.
+    ///
+    /// Returns [`None`] if not found. If the second parameter is present, it
+    /// specifies the position in the string to begin the search.
```
Is it "byte position" or "character position"?

If it is byte index, can you add a test (not a doc test, a real one) where the position is in the middle of a multi byte UTF-8 character?

choznerol replied:
> Is it "byte position" or "character position"?

fixed in d8eda7b7fe5eba53c460b4d80c28694da2905fe6

> If it is byte index, can you add a test (not a doc test, a real one) where the position is in the middle of a multi byte UTF-8 character?

Sure. I test this as part of `byteindex_support_specifiying_byte_position_to_start_search`, added in 0351cb2.

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1684,6 +1705,89 @@ impl String {
+    /// given substring.
+    ///
+    /// Returns [`None`] if not found. If the second parameter is present, it
+    /// specifies the position in the string to begin the search.
```
Ditto. Can we clarify byte ha char position?

choznerol replied:
fixed in d8eda7b7fe5eba53c460b4d80c28694da2905fe6

lopopolo commented on spinoso-string/src/lib.rs:
````diff
@@ -1684,6 +1705,89 @@ impl String {
+    /// assert_eq!(s.byteindex("X", None), None);
+    /// ```
+    ///
+    /// [`String#index`]: https://ruby-doc.org/3.2.0/String.html#method-i-byteindex
````
Markdown Doc link anchor needs to be updated

choznerol replied:
also fixed in 2b7662f0d0d1c4bf695e9f2262af61f17da26f22

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1684,6 +1705,89 @@ impl String {
+    /// given substring.
+    ///
+    /// Returns [`None`] if not found. If the second parameter is present, it
+    /// specifies the position in the string to begin the search.
```
Ditto. Clarify byte vs char pos

choznerol replied:
also fixed in d8eda7b7fe5eba53c460b4d80c28694da2905fe6

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1684,6 +1705,89 @@ impl String {
         inner(self.inner.as_slice(), needle, offset)
     }
 
+    /// Returns the Integer byte-based index of the first occurrence of the
```
Can we make this match the first line of the docs for the rust functions `index` and `rindex`. The talk of "Integer" I think comes straight from the Ruby docs but is a bit confusing in a Rust doc comment context.

choznerol replied:
also fixed in d8eda7b7fe5eba53c460b4d80c28694da2905fe6

lopopolo reviewed (changes requested):
This change looks good, the functions do what I expect. Thanks for updating the doc tests to have multibyte UTF-8 characters.

I think you've got about the right amount of doc test examples.

Can you add real tests that are a bit more exhaustive? I'd like to see the following coverage:

- all encodings for the receiver
- `None` given as needle
- `Some([])`
- ASCII needles
- Binary needles
- UTF-8 needles with multibyte chars
- testing on UTF-8 encoded strings that have binary contents.

choznerol committed:
Fix typo

https://github.com/artichoke/artichoke/pull/2449#discussion_r1144081120
https://github.com/artichoke/artichoke/pull/2449#discussion_r1144083300

choznerol committed:
Update comments for be specific about char v.s. byte

https://github.com/artichoke/artichoke/pull/2449#discussion_r1144084915
https://github.com/artichoke/artichoke/pull/2449#discussion_r1144082458
https://github.com/artichoke/artichoke/pull/2449#discussion_r1144082831
https://github.com/artichoke/artichoke/pull/2449#discussion_r1144084249

choznerol committed:
Prefer String::utf8 over String::from for clarity

https://github.com/artichoke/artichoke/pull/2449#discussion_r1144081467
https://github.com/artichoke/artichoke/pull/2449#discussion_r1144081685

choznerol committed:
Test String::byteindex() with needle and heystack of different encodings

https://github.com/artichoke/artichoke/pull/2449#pullrequestreview-1351515379

lopopolo reviewed (approved):
this looks great.

Once https://github.com/rust-lang/rust/pull/109500 is resolved I think the rustdoc build will be fixed and I'll poke CI to get this merged.

choznerol commented:
(Ops sorry I somehow didn't send out the comments yesterday 😅)

> Can you add real tests that are a bit more exhaustive? I'd like to see the following coverage:
> 
>     * all encodings for the receiver
>     * `None` given as needle
>     * `Some([])`
>     * ASCII needles
>     * Binary needles
>     * UTF-8 needles with multibyte chars
>     * testing on UTF-8 encoded strings that have binary contents.

Thanks for the detailed review! I've added the test only for `byteindex` in d9ce3c8, but not for `byterindex` to avoid duplication.

choznerol committed:
Test 'position' parameter of byteindex

lopopolo reviewed (approved):
Thank you this is good to go. It looks like the rust doc folks are making progress on the nightly breakage. I'm following upstream and will poke the build here when nightly is fixed so we can get this merged.

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -2117,4 +2221,56 @@ mod tests {
+
+        // Empty string as needle
+        let needle = String::utf8("a".as_bytes().to_vec());
+        assert_eq!(utf8.byteindex(needle.clone(), None), Some(0));
```
It looks like clippy isn't happy with all of these clones.

`spinoso_string::String` impls `AsRef<[u8]>` so these `needle.clone()` can be replaced with `&needle`.

choznerol replied:
Oops, thanks I didn't notice the warning. Fixed in 973224d.

choznerol committed:
Fix clippy warnings

https://github.com/artichoke/artichoke/pull/2449#discussion_r1161054722

lopopolo commented:
nightly rustdoc is fixed, the build is green, merging. thanks for the contribution @choznerol.
